README.md
# MoasdaWiki Server

## Description

MoasdaWiki Server is a privacy-friendly and interactive knowledge management
tool. It provides a browser-based GUI to search, create, and modify content in
Wiki style.

For documentation see https://moasdawiki.net/.

### Key features

- Easy usability
- No installation required, can be unzipped and run on a USB stick
- Runs on any OS with a Java VM 1.8 or higher
- Debian package for easy installation available
- [MoasdaWiki App](https://gitlab.com/moasdawiki/moasdawiki-app) available,
  always have a copy of the Wiki content on your mobile device
- Data privacy by design: Stores content only on the local computer, never
  establishes a cloud connection
- Free/Libre and Open Source Software (FLOSS)

### Feature list

- Powerful syntax, supports tables and images
- Fast full-text search
- Page templates to create uniform looking pages
- Full customizable graphical user interface
  - Navigation panel can be modified
  - Page layout for view in browser and for printing is defined by CSS stylesheet
  - Page header and footer can be modified
  - Customizable HTML header, e.g. to include your own CSS stylesheets and JavaScript libraries
  - Configuration file with several options
- Internationalization via message files
- List of recently modified and viewed pages
- Built-in lightweight web server, no external web server required
- Stores pages as text files in a folder in the local file system;
  for backups just make a copy of that folder.

## Screenshots

<img src="images/Screenshot1.png" width="300" height="256" />
<img src="images/Screenshot2.png" width="300" height="256" />
<img src="images/Screenshot3.png" width="300" height="256" />

## Download and Run
### For Debian/Ubuntu

1. Download the latest DEB package file from the
   [releases page](https://gitlab.com/moasdawiki/moasdawiki-server/-/releases).
2. Install the package.
3. Open the URL http://localhost:11080/ in your web browser.

MoasdaWiki stores the repository files in the folder
`/var/lib/moasdawiki/repository`. I recommend adding that folder to your backup
tool. The log files are in the folder `/var/log/moasdawiki/`.

### Other environments

1. Ensure that Java 1.8 or higher is installed.
2. Download the latest ZIP bundle from the
   [releases page](https://gitlab.com/moasdawiki/moasdawiki-server/-/releases).
3. Unzip the file `moasdawiki-server-x.y.z.zip` in a new folder.
4. Open a Terminal.
5. Go to the folder that contains the file `moasdawiki-server-x.y.z.jar`.
6. Run the server: `java -jar moasdawiki-server-x.y.z.jar shadow-repository/en`
7. Open the URL http://localhost:11080/ in your web browser.

If the Wiki server runs on a different host, use the host name or its IP address
instead of `localhost`. The default port is 11080. You can change it in the file
`shadow-repository/en/config.txt`.

## Build

Clone the GitLab repository and run:

```
gradle build
```

Afterwards, you will find a ZIP file in the folder ```build/distributions/```.

## User guide

A detailed description of the MoasdaWiki server is bundled in the downloaded ZIP file.
To open the user guide click on the "Help" link in the navigation panel or open the URL
http://localhost:11080/view/wiki/

## Support

If you have questions or any problems you can contact me via [support@moasdawiki.net]().

## License

MoasdaWiki server is licensed under AGPL-3.0-only &ndash; see the
[LICENSE](LICENSE) file for details.

Copyright (C) Herbert Reiter


CHANGELOG.md
# Changelog

## 3.0.0 (2021-07-03)

- Create Debian package (DEB) for easier installation and update
- Implement shadow repository to separate shipped repository files from user files

## 2.6.3 (2021-06-15)

- Syntax highlighting for YAML code
- Extend Java syntax highlighting for annotations
- Improve XML and HTML syntax highlighting

## 2.6.2 (2021-06-13)

- Syntax highlighting for ini code
- Upgrade build dependencies

## 2.6.1 (2021-05-22)

- Change License to AGPL-3.0-only

## 2.6.0 (2021-05-13)

- Make GPL-3.0-only licensing more clear

## 2.5.3 (2021-05-11)

- More strict XML parsing
- Small code improvements

## 2.5.2 (2021-02-28)

- Find also words that start with the search string

## 2.5.1 (2021-01-31)

- Bugfix: Add wiki page name to search index

## 2.5.0 (2021-01-31)

- Enhance search index to full words
- Lazy load search index
- Introduce search ignore list
- Increase file upload size to 10 MB
- Add support for SVG images
- Refactoring

## 2.4.4 (2021-01-18)

- Search results: Don't scan wiki page content in App to speed up search
- Search results: Sort by relevance and page name
- Search results: Hide relevance points

## 2.4.3 (2021-01-04)

- Update README.md
- Code changes for App support
- Bugfix: Reset enumeration counting
- Fix primitive array annotation warnings
- Update copyright year

## 2.4.2 (2020-12-28)

- Refactoring cache updates for App support
- Optimize log output

## 2.4.1 (2020-12-27)

- Small changes for App support
- Bug fixes

## 2.4.0 (2020-12-26)

- Internationalize wiki editor and synchronization page
- Upgrade to HTML 5
- HTML pages specify language code
- Refactoring: Remove plugin support
- Refactoring: Remove RepositoryService interface
- Refactoring: Remove WikiService interface
- Refactoring: Remove WikiService file content cache
- Clean up config file
- Upgrade build dependencies
- Update JavaDoc comments

## 2.3.3 (2020-10-11)

- Add SearchService read-only mode for App

## 2.3.2 (2020-10-11)

- Extend wiki search to find also numbers
- Small code changes

## 2.3.1 (2020-06-21)

- Extend wiki search to find also upper case words with 3 characters

## 2.3.0 (2020-06-14)

- Speed up full text search by a search index
- Do not store last synchronization timestamp in session list to avoid file modification on every synchronization attempt
- Bugfix: Exclude session list from synchronization with App
- Refactorings

## 2.2.0 (2020-02-08)

- Speed up full text search by searching only in the page titles in the first step
- Add unit tests
- Refactorings

## 2.1.0 (2020-01-04)

- Code refactored to use the new language features of Java 1.8
- Replaced JAXB implementation by my own one to keep it running on future Java releases
- Added `@Nullable` and `@NotNull` annotations, fixed potential bugs
- RepositoryService and WikiService refactored, rewrote parts of the implementation
- Removed UDP broadcast support as it didn't really work for the app 
- Moved app sources to a separate project
- Updated copyright
- Ant build script replaced by Gradle, restructured repository folders
- Published sources on GitLab
- Using GitLab CI

## 2.0.9 (2019-01-03)

- HTML layout changed to border-box
- App: targetSdkVersion updated due to recent Google requirements
- App: Doesn't use an internal TCP port any more. Thus, the app requires no Internet permission for normal usage, it is only necessary for synchronisation.
- App (versionCode 10)
- App (versionCode 11)
- App (versionCode 12)
  - App: Bugfix: Fixed broken search
- App 2.0.9.2 (versionCode 13)
  - App: Bugfix: Page titles with space or special characters couldn't be shown

## 2.0.8 (2018-02-05)

- App can be closed by pressing the back button twice.
- Bugfix: Children pages were not restored after editing a wiki page.
- App (versionCode 9)

## 2.0.7 (2017-05-28)

- Bugfix: App: Editor icons got visible on scrolling
- App (versionCode 7)

## 2.0.6 (2016-11-28)

- Bugfix: JavaScript was not loaded in English version
- Bugfix: Some missing translations added
- Bugfix: App: Edit symbol was still available and active
- App (versionCode 6)

## 2.0.5 (2016-11-22)

- Initial App version (versionCode 5)

## 2.0.1 (2016-11-22)

- Bugfix: Error in path handling in Windows, media files were not found

## 2.0.0 (2016-11-21)

- MoasdaWiki as Android App (versionCode 2)
  - App provides a copy for mobile usage
  - Synchronizes with a MoasdaWiki server
  - Optimized for small display resolution, uses integrated navigation of app
- New syntax element class WikiTag
- New plugin for synchronizing with the app
  - Status view and grant app access via the synchronization wiki page
  - UDP broadcast search by app to find the server
- Repository access supports both, app assets and Android Internal Storage
- Added cache files to speed up app startup
- New GUI design
- Bugfix: Firefox requests sometimes got stuck for some seconds (speculative requests).

## 1.12 (2015-01-18)

- `@@` tag now supports syntax highlighting for Java, HTML, XML, and Properties
- Rendering of multi line code sections (`@@...@@`) improved to support copy & paste preserving line breaks
- Formatting of date and time now in English format, can be changed by message file
- Source code now requires Java 1.7, updated documentation
- Bugfix: Image URLs were not escaped correctly, e.g. if URL contains the character `+`
- Bugfix: Index pages did not show their last modification timestamp

## 1.11 (2014-06-21)

- Translated project into English
- Bugfix: Page links with special character `!` didn't work
- Bugfix: Wiki editor had problems with special characters in page name


src/main/java/net/moasdawiki/base/Settings.java
/*
 * MoasdaWiki Server
 *
 * Copyright (C) 2008 - 2021 Herbert Reiter (herbert@moasdawiki.net)
 *
 * This program is free software: you can redistribute it and/or modify it
 * under the terms of the GNU Affero General Public License version 3 as
 * published by the Free Software Foundation (AGPL-3.0-only).
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see
 * <https://www.gnu.org/licenses/agpl-3.0.html>.
 */

package net.moasdawiki.base;

import net.moasdawiki.service.repository.AnyFile;
import net.moasdawiki.service.repository.RepositoryService;
import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.BufferedReader;
import java.io.StringReader;
import java.net.InetAddress;
import java.net.UnknownHostException;
import java.util.*;

/**
 * Reads the settings file and provides methods to get settings. Multiple
 * settings with the same key are combined to a list value.
 */
@SuppressWarnings("SameReturnValue")
public class Settings {

	private static final String PROGRAM_NAME = "MoasdaWiki";
	private static final String CONFIG_FILE_SERVER = "/config.txt";
	private static final String CONFIG_FILE_APP = "/config-app.txt";

	private static final String SERVERPORT = "port";
	private static final int SERVERPORT_DEFAULT = 11080;
	private static final String MESSAGE_FILE = "messagefile";
	private static final String MESSAGE_FILE_DEFAULT = "/wiki/messages.txt";
	private static final String PAGE_STARTPAGE = "page.startpage";
	private static final String PAGE_STARTPAGE_DEFAULT = "/Home";
	private static final String PAGE_NAVIGATION = "page.navigation";
	private static final String PAGE_NAVIGATION_DEFAULT = "/wiki/Navigation";
	private static final String PAGE_HTML_HEADER = "page.html.header";
	private static final String PAGE_HTML_HEADER_DEFAULT = "/wiki/HtmlHeader";
	private static final String PAGE_HEADER = "page.header";
	private static final String PAGE_HEADER_DEFAULT = "/wiki/PageHeader";
	private static final String PAGE_FOOTER = "page.footer";
	private static final String PAGE_FOOTER_DEFAULT = "/wiki/PageFooter";
	private static final String PAGE_TEMPLATES = "page.templates";
	private static final String PAGE_TEMPLATES_DEFAULT = "/wiki/Templates";
	private static final String PAGE_INDEX_NAME = "page.index.name";
	private static final String PAGE_INDEX_NAME_DEFAULT = "Index";
	private static final String PAGE_INDEX_GLOBAL = "page.index.default";
	private static final String PAGE_INDEX_GLOBAL_DEFAULT = "/wiki/IndexStandard";
	private static final String REPOSITORY_ROOT = "repository.root";
	private static final String REPOSITORY_ROOT_DEFAULT = "root";
	private static final String AUTHENTICATION_ONLYLOCALHOST = "authentication.onlylocalhost";

	@NotNull
	private final Logger logger;

	/**
	 * Contains all settings. The map value is either a string or a list of strings.
	 */
	@NotNull
	private final Map<String, Object> settings;

	@NotNull
	private final RepositoryService repositoryService;

	@NotNull
	private final String configFileName;

	/**
	 * Constructor.
	 */
	public Settings(@NotNull Logger logger, @NotNull RepositoryService repositoryService, @NotNull String configFileName) {
		this.logger = logger;
		this.settings = new HashMap<>();
		this.repositoryService = repositoryService;
		this.configFileName = configFileName;
		reset();
	}

	/**
	 * Rereads the configuration file.
	 * Is called in App environment after synchronization with server.
	 */
	public void reset() {
		try {
			AnyFile anyFile = new AnyFile(configFileName);
			String settingsContent = repositoryService.readTextFile(anyFile);
			BufferedReader reader = new BufferedReader(new StringReader(settingsContent));
			String line;
			while ((line = reader.readLine()) != null) {
				extractMapping(line);
			}
			reader.close();
		} catch (Exception e) {
			logger.write("Cannot read settings file: " + e.getMessage());
		}
	}

	/**
	 * Parses a single line of the configuration file.
	 */
	private void extractMapping(@NotNull String line) {
		if (line.startsWith("#") || line.startsWith("//")) {
			return; // ignore line comment
		}

		int pos = line.indexOf('=');
		if (pos == -1) {
			// no valid key-value-pair with '=' delimiter
			return;
		}

		// extract key and value
		String key = line.substring(0, pos).trim();
		String value = line.substring(pos + 1).trim();

		// add to settings map
		Object previousValue = settings.get(key);
		if (previousValue instanceof String) {
			// key is already mapped to a String value
			// --> create String list and add second value
			List<String> values = new ArrayList<>();
			values.add((String) previousValue);
			values.add(value);
			settings.put(key, values);
		} else if (previousValue instanceof List<?>) {
			// add value to existing String list
			@SuppressWarnings("unchecked")
			List<String> values = (List<String>) previousValue;
			values.add(value);
		} else {
			// first assignment of key
			settings.put(key, value);
		}
	}

	/**
	 * Returns a string value.
	 * 
	 * @param key Setting key, not <code>null</code>.
	 * @return Setting value. <code>null</code> --> no setting for the key.
	 */
	@Nullable
	public String getString(@NotNull String key) {
		return getString(key, null);
	}

	/**
	 * Returns a string value. If the key has no value, the given default value
	 * is returned instead.
	 * 
	 * @param key Setting key, not <code>null</code>.
	 * @param defaultValue Default value.
	 * @return Setting value or default value. Not <code>null</code> if default
	 *         value is not <code>null</code>.
	 */
	@Nullable
	@Contract(value = "_, !null -> !null", pure = true)
	public String getString(@NotNull String key, @Nullable String defaultValue) {
		Object value = settings.get(key);
		if (value instanceof String) {
			return (String) value;
		}
		if (value instanceof List<?>) {
			List<?> values = (List<?>) value;
			if (values.size() >= 1) {
				return (String) values.get(0);
			}
		}
		return defaultValue;
	}

	/**
	 * Returns a string array value.
	 * 
	 * @param key Setting key, not <code>null</code>.
	 * @param defaultValues Default values.
	 * @return Setting value or default value. Not <code>null</code>.
	 */
	@NotNull
	public String[] getStringArray(@NotNull String key, String... defaultValues) {
		Object value = settings.get(key);
		if (value instanceof List<?>) {
			List<?> valueList = (List<?>) value;
			if (!valueList.isEmpty() && valueList.get(0) instanceof String) {
				//noinspection SuspiciousToArrayCall
				return valueList.toArray(new String[0]);
			} else {
				return new String[0];
			}
		} else if (value instanceof String) {
			String[] result = new String[1];
			result[0] = (String) value;
			return result;
		}
		return defaultValues;
	}

	/**
	 * Returns an int value.
	 * 
	 * @param key Setting key, not <code>null</code>.
	 * @param defaultValue Default value.
	 * @return Setting value or default value.
	 */
	@SuppressWarnings("unused")
	public int getInt(@NotNull String key, int defaultValue) {
		return getInteger(key, defaultValue);
	}

	/**
	 * Returns an integer value.
	 * 
	 * @param key Setting key, not <code>null</code>.
	 * @param defaultValue Default value. <code>null</code> --> no default
	 *        value.
	 * @return Setting value or default value. <code>null</code> --> setting not
	 *         existing or not an integer.
	 */
	@Nullable
	@Contract(value = "_, !null -> !null")
	public Integer getInteger(@NotNull String key, @Nullable Integer defaultValue) {
		Integer value = defaultValue;
		String strValue = getString(key);
		try {
			if (strValue != null) {
				value = Integer.parseInt(strValue);
			}
		} catch (NumberFormatException e) {
			logger.write("Setting value for '" + key + "' is not a number: '" + strValue + "'");
		}
		return value;
	}

	/**
	 * Returns a boolean value.
	 * 
	 * @param key Setting key, not <code>null</code>.
	 * @param defaultValue Default value.
	 * @return Setting value or default value.
	 */
	public boolean getBoolean(@NotNull String key, boolean defaultValue) {
		boolean value = defaultValue;
		String strValue = getString(key);
		if ("true".equalsIgnoreCase(strValue)) {
			value = true;
		} else if ("false".equalsIgnoreCase(strValue)) {
			value = false;
		} else {
			logger.write("Setting value for '" + key + "' is not true or false: '" + strValue + "'");
		}
		return value;
	}

	/**
	 * Returns the program name.
	 */
	@NotNull
	public String getProgramName() {
		return PROGRAM_NAME;
	}

	/**
	 * Returns the version number.
	 */
	@NotNull
	public String getVersion() {
		Package aPackage = Settings.class.getPackage();
		String implementationVersion = aPackage.getImplementationVersion();
		if (implementationVersion != null) {
			return implementationVersion;
		} else {
			return "SNAPSHOT";
		}
	}

	/**
	 * Returns the program name inclusive version number.
	 */
	@NotNull
	public String getProgramNameVersion() {
		return getProgramName() + " " + getVersion();
	}

	/**
	 * Returns the file path of the message file.
	 */
	@NotNull
	public String getMessageFile() {
		return getString(MESSAGE_FILE, MESSAGE_FILE_DEFAULT);
	}

	/**
	 * Returns the host name of the wiki server. <code>null</code> --> unknown.
	 */
	@Nullable
	public String getServerHost() {
		try {
			return InetAddress.getLocalHost().getHostName();
		} catch (UnknownHostException e) {
			return null;
		}
	}

	/**
	 * Returns the port of the wiki server.
	 */
	public int getServerPort() {
		return getInteger(SERVERPORT, SERVERPORT_DEFAULT);
	}

	/**
	 * Wiki page to be shown as start page.
	 */
	@NotNull
	public String getStartpagePath() {
		return getString(PAGE_STARTPAGE, PAGE_STARTPAGE_DEFAULT);
	}

	/**
	 * Wiki page to be shown in navigation area. <code>null</code> --> no
	 * navigation area.
	 */
	@Nullable
	public String getNavigationPagePath() {
		return getString(PAGE_NAVIGATION, PAGE_NAVIGATION_DEFAULT);
	}

	/**
	 * Wiki page to be included as HTML header lines. <code>null</code> --> no
	 * additional HTML header lines.
	 */
	@Nullable
	public String getHtmlHeaderPagePath() {
		return getString(PAGE_HTML_HEADER, PAGE_HTML_HEADER_DEFAULT);
	}

	/**
	 * Wiki page to be shown on top of all wiki pages. <code>null</code> --> no
	 * page header.
	 */
	@Nullable
	public String getHeaderPagePath() {
		return getString(PAGE_HEADER, PAGE_HEADER_DEFAULT);
	}

	/**
	 * Wiki page to be shown on bottom of all wiki pages. <code>null</code> -->
	 * no page footer.
	 */
	@Nullable
	public String getFooterPagePath() {
		return getString(PAGE_FOOTER, PAGE_FOOTER_DEFAULT);
	}

	/**
	 * Parent wiki page of content templates.
	 */
	@Nullable
	public String getTemplatesPagePath() {
		return getString(PAGE_TEMPLATES, PAGE_TEMPLATES_DEFAULT);
	}

	/**
	 * Page name that is used as index (summary) of a repository folder
	 * if the URL path ends with '/'.
	 * null -> Disable index feature.
	 */
	@Nullable
	public String getIndexPageName() {
		return getString(PAGE_INDEX_NAME, PAGE_INDEX_NAME_DEFAULT);
	}

	/**
	 * Global index page to be used as fallback if a repository folder has no
	 * own index page (see {@link #getIndexPageName()}).
	 */
	@Nullable
	public String getIndexFallbackPagePath() {
		return getString(PAGE_INDEX_GLOBAL, PAGE_INDEX_GLOBAL_DEFAULT);
	}

	/**
	 * Repository subfolder that contains the files that are accessible via URL
	 * root path, e.g. favicon.ico.
	 */
	@NotNull
	public String getRootPath() {
		return getString(REPOSITORY_ROOT, REPOSITORY_ROOT_DEFAULT);
	}

	/**
	 * Should the wiki server only be accessible for clients running on local
	 * host?
	 */
	public boolean isOnlyLocalhostAccess() {
		return getBoolean(AUTHENTICATION_ONLYLOCALHOST, false);
	}

	/**
	 * Returns the current system time.
	 */
	@NotNull
	public Date getActualTime() {
		return new Date();
	}

	@NotNull
	public static String getConfigFileServer() {
		return CONFIG_FILE_SERVER;
	}

	@SuppressWarnings("unused")
	@NotNull
	public static String getConfigFileApp() {
		return CONFIG_FILE_APP;
	}
}
